libcontainer/factory_linux.go

```diff
@@ -239,19 +239,23 @@ func (l *LinuxFactory) StartInitialization() (err error) {
 		}
 		*pair.v = i
 	}
+
 	var (
 		pipe = os.NewFile(uintptr(pipefd), "pipe")
 		it   = initType(os.Getenv("_LIBCONTAINER_INITTYPE"))
 	)
+
+	// ensure that this pipe is always closed in the case of failure.
+	defer pipe.Close()
```

Comment:
When I said "immediately" I meant inside the function. I understand what you're saying, but especially because of `recover()` I don't like moving this outside of the other `defer`. `EOF`s are important inside runC, and I don't like splitting that from where the error handling is done.